src/app/shared/services/app-config/app-config.service.ts

```diff
@@ -36,35 +38,52 @@ export class AppConfigService extends SyncServiceBase {
     return this.appConfigChanges$;
   }
 
-  /** Track deep object diff of app config changes, including full initial value */
+  /**
+   * @deprecated - prefer use of config signal and computed/effect bindings
+   * Track deep object diff of app config changes, including full initial value
+   * */
   public get changesWithInitialValue$() {
-    return this.changes$.pipe(startWith(this.value));
+    return this.changes$.pipe(startWith(this.appConfig()));
   }
 
-  constructor(private deploymentService: DeploymentService) {
+  constructor(
+    private deploymentService: DeploymentService,
+    private router: Router
+  ) {
     super("AppConfig");
     this.initialise();
   }
 
+  /** When service initialises load any deployment-specific config overrides  */
   private initialise() {
-    const deploymentOverrides: IAppConfigOverride = this.deploymentService.config.app_config || {};
-    this.APP_CONFIG = getDefaultAppConfig();
-    // Store app config with deployment overrides applied, to be merged with additional overrides when applied
-    this.deploymentAppConfig = this.applyAppConfigOverrides(this.APP_CONFIG, deploymentOverrides);
-    this.updateAppConfig(deploymentOverrides);
+    // When first loading handle side-effects from default config (e.g. initial routing).
+    // Deployment-specific side-effects will be handled when setting the appConfig
+    const defaultConfig = getDefaultAppConfig();
+    this.handleConfigSideEffects(defaultConfig, defaultConfig);
+
+    // Set app config using deployment overrides
+    this.setAppConfig(this.deploymentService.config.app_config);
   }
 
-  public updateAppConfig(overrides: IAppConfigOverride) {
-    // Clone this.deploymentAppConfig so that the original is unaffected by deepMergeObjects()
-    const appConfigWithOverrides = this.applyAppConfigOverrides(
-      clone(this.deploymentAppConfig),
-      overrides
-    );
-    this.APP_CONFIG = appConfigWithOverrides;
-    this.appConfig$.next(appConfigWithOverrides);
+  /**
+   * Generate a complete app config by deep-merging app config overrides
+   * with the default config
+   * @param overrides
+   * @returns
+   */
+  public setAppConfig(overrides: IAppConfigOverride = {}) {
+    // Ignore case where no overrides provides or overrides already applied
+    if (Object.keys(overrides).length === 0) return;
+
+    const mergedConfig = deepMergeObjects(getDefaultAppConfig(), overrides);
```

Comment:
Ah good point, yeah I was a bit confused on this when working on it. I think some of the confusion was we were using an `updateAppConfig` variable but the update function was still pulling the default config and merging updates on top (so not so much updating the current config but setting a new one from a partial config)

So if I understand skins correctly, they are a layer on top of the compiled deployment config. In the case of applying successive skins, the 1st skin is simply a patch on the default deployment, but the 2nd would need to revert changes from the first and then apply. Or more simply each skin will want to independently set a complete value based on merge with default config.

I'll update the skin service to handle the partial merging part so that the appConfig methods can stay more straight-forward

Reply:
So it took quite a bit more work than expected but I think I've just about got there in the end. I realised one of the issues with the way app skins were currently being applied was that each time the updated skin config was merged on top of the original deployment config, meaning that if any app config had been updated from anywhere else during that time the changes would be lost.

In practice this wasn't a huge issue as the only other service that updated directly was the app header (and only during testing/debug of scroll collapse feature). Although definitely all the different levels of config changes were quite confusing and difficult to unpick.

In the end I found myself needing to write test suites to properly understand the behaviour and iterate on the code. The way the skin system now (should) work is that each time a skin is applied it also generates a `revert` override which can be used to undo the skin-specific changes. Then if the skin is changed again the new skin update can be merged with the previous revert to leave an overall config that would be the same had the second skin been applied originally, but allowing for any independent config changes that may have occurred in that time. 

If you get a chance could you take a look at the test cases and see if they make sense to you. I've manually checked toggling between the two skins on the `debug_skin_select` template which works, although might be good to check against any production deployment that also uses (I'm not sure which do)

I also separately realised that various recent refactors had broken some of the service spec tests (which we don't run as part of CI (outstanding #2196), so I also have fixed these

Reply:
Unfortunately the updates are spread across a number of commits, but all the meaningful changes should be within the `skin.service.ts` and `skin.service.spec.ts` files

Reply:
See my latest review comment for a continuation of this discussion, as well as https://github.com/IDEMSInternational/open-app-builder/pull/2423#discussion_r1765259530